In [1]:
import pandas as pd

# fixing the Date datatype

In [3]:
df=pd.read_csv("Coffe_sales.csv",parse_dates=["Date"])
df

hour_of_day cash_type  money    coffee_name Time_of_Day Weekday  \
0              10      card  38.70          Latte     Morning     Fri   
1              12      card  38.70  Hot Chocolate   Afternoon     Fri   
2              12      card  38.70  Hot Chocolate   Afternoon     Fri   
3              13      card  28.90      Americano   Afternoon     Fri   
4              13      card  38.70          Latte   Afternoon     Fri   
...           ...       ...    ...            ...         ...     ...   
3542           10      card  35.76     Cappuccino     Morning     Sun   
3543           14      card  35.76          Cocoa   Afternoon     Sun   
3544           14      card  35.76          Cocoa   Afternoon     Sun   
3545           15      card  25.96      Americano   Afternoon     Sun   
3546           18      card  35.76          Latte       Night     Sun   

     Month_name  Weekdaysort  Monthsort       Date             Time  
0           Mar            5          3 2024-03-01  10:15:50.520000  
1           Mar            5          3 2024-03-01  12:19:22.539000  
2           Mar            5          3 2024-03-01  12:20:18.089000  
3           Mar            5          3 2024-03-01  13:46:33.006000  
4           Mar            5          3 2024-03-01  13:48:14.626000  
...         ...          ...        ...        ...              ...  
3542        Mar            7          3 2025-03-23  10:34:54.894000  
3543        Mar            7          3 2025-03-23  14:43:37.362000  
3544        Mar            7          3 2025-03-23  14:44:16.864000  
3545        Mar            7          3 2025-03-23  15:47:28.723000  
3546        Mar            7          3 2025-03-23  18:11:38.635000  

[3547 rows x 11 columns]

In [4]:
df.head()

hour_of_day cash_type  money    coffee_name Time_of_Day Weekday Month_name  \
0           10      card   38.7          Latte     Morning     Fri        Mar   
1           12      card   38.7  Hot Chocolate   Afternoon     Fri        Mar   
2           12      card   38.7  Hot Chocolate   Afternoon     Fri        Mar   
3           13      card   28.9      Americano   Afternoon     Fri        Mar   
4           13      card   38.7          Latte   Afternoon     Fri        Mar   

   Weekdaysort  Monthsort       Date             Time  
0            5          3 2024-03-01  10:15:50.520000  
1            5          3 2024-03-01  12:19:22.539000  
2            5          3 2024-03-01  12:20:18.089000  
3            5          3 2024-03-01  13:46:33.006000  
4            5          3 2024-03-01  13:48:14.626000

In [5]:
df.tail()

hour_of_day cash_type  money coffee_name Time_of_Day Weekday Month_name  \
3542           10      card  35.76  Cappuccino     Morning     Sun        Mar   
3543           14      card  35.76       Cocoa   Afternoon     Sun        Mar   
3544           14      card  35.76       Cocoa   Afternoon     Sun        Mar   
3545           15      card  25.96   Americano   Afternoon     Sun        Mar   
3546           18      card  35.76       Latte       Night     Sun        Mar   

      Weekdaysort  Monthsort       Date             Time  
3542            7          3 2025-03-23  10:34:54.894000  
3543            7          3 2025-03-23  14:43:37.362000  
3544            7          3 2025-03-23  14:44:16.864000  
3545            7          3 2025-03-23  15:47:28.723000  
3546            7          3 2025-03-23  18:11:38.635000

In [6]:
df.head(10)

hour_of_day cash_type  money          coffee_name Time_of_Day Weekday  \
0           10      card   38.7                Latte     Morning     Fri   
1           12      card   38.7        Hot Chocolate   Afternoon     Fri   
2           12      card   38.7        Hot Chocolate   Afternoon     Fri   
3           13      card   28.9            Americano   Afternoon     Fri   
4           13      card   38.7                Latte   Afternoon     Fri   
5           15      card   33.8  Americano with Milk   Afternoon     Fri   
6           16      card   38.7        Hot Chocolate   Afternoon     Fri   
7           18      card   33.8  Americano with Milk       Night     Fri   
8           19      card   38.7                Cocoa       Night     Fri   
9           19      card   33.8  Americano with Milk       Night     Fri   

  Month_name  Weekdaysort  Monthsort       Date             Time  
0        Mar            5          3 2024-03-01  10:15:50.520000  
1        Mar            5          3 2024-03-01  12:19:22.539000  
2        Mar            5          3 2024-03-01  12:20:18.089000  
3        Mar            5          3 2024-03-01  13:46:33.006000  
4        Mar            5          3 2024-03-01  13:48:14.626000  
5        Mar            5          3 2024-03-01  15:39:47.726000  
6        Mar            5          3 2024-03-01  16:19:02.756000  
7        Mar            5          3 2024-03-01  18:39:03.580000  
8        Mar            5          3 2024-03-01  19:22:01.762000  
9        Mar            5          3 2024-03-01  19:23:15.887000

In [7]:
df.tail(10)

hour_of_day cash_type  money          coffee_name Time_of_Day Weekday  \
3537           15      card  21.06             Espresso   Afternoon     Sat   
3538           16      card  35.76                Latte   Afternoon     Sat   
3539           17      card  30.86  Americano with Milk       Night     Sat   
3540           19      card  30.86  Americano with Milk       Night     Sat   
3541           10      card  35.76                Latte     Morning     Sun   
3542           10      card  35.76           Cappuccino     Morning     Sun   
3543           14      card  35.76                Cocoa   Afternoon     Sun   
3544           14      card  35.76                Cocoa   Afternoon     Sun   
3545           15      card  25.96            Americano   Afternoon     Sun   
3546           18      card  35.76                Latte       Night     Sun   

     Month_name  Weekdaysort  Monthsort       Date             Time  
3537        Mar            6          3 2025-03-22  15:57:58.183000  
3538        Mar            6          3 2025-03-22  16:26:59.766000  
3539        Mar            6          3 2025-03-22  17:53:35.942000  
3540        Mar            6          3 2025-03-22  19:16:52.727000  
3541        Mar            7          3 2025-03-23  10:07:11.278000  
3542        Mar            7          3 2025-03-23  10:34:54.894000  
3543        Mar            7          3 2025-03-23  14:43:37.362000  
3544        Mar            7          3 2025-03-23  14:44:16.864000  
3545        Mar            7          3 2025-03-23  15:47:28.723000  
3546        Mar            7          3 2025-03-23  18:11:38.635000

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3547 entries, 0 to 3546
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   hour_of_day  3547 non-null   int64         
 1   cash_type    3547 non-null   object        
 2   money        3547 non-null   float64       
 3   coffee_name  3547 non-null   object        
 4   Time_of_Day  3547 non-null   object        
 5   Weekday      3547 non-null   object        
 6   Month_name   3547 non-null   object        
 7   Weekdaysort  3547 non-null   int64         
 8   Monthsort    3547 non-null   int64         
 9   Date         3547 non-null   datetime64[ns]
 10  Time         3547 non-null   object        
dtypes: datetime64[ns](1), float64(1), int64(3), object(6)
memory usage: 304.9+ KB


In [9]:
df

hour_of_day cash_type  money    coffee_name Time_of_Day Weekday  \
0              10      card  38.70          Latte     Morning     Fri   
1              12      card  38.70  Hot Chocolate   Afternoon     Fri   
2              12      card  38.70  Hot Chocolate   Afternoon     Fri   
3              13      card  28.90      Americano   Afternoon     Fri   
4              13      card  38.70          Latte   Afternoon     Fri   
...           ...       ...    ...            ...         ...     ...   
3542           10      card  35.76     Cappuccino     Morning     Sun   
3543           14      card  35.76          Cocoa   Afternoon     Sun   
3544           14      card  35.76          Cocoa   Afternoon     Sun   
3545           15      card  25.96      Americano   Afternoon     Sun   
3546           18      card  35.76          Latte       Night     Sun   

     Month_name  Weekdaysort  Monthsort       Date             Time  
0           Mar            5          3 2024-03-01  10:15:50.520000  
1           Mar            5          3 2024-03-01  12:19:22.539000  
2           Mar            5          3 2024-03-01  12:20:18.089000  
3           Mar            5          3 2024-03-01  13:46:33.006000  
4           Mar            5          3 2024-03-01  13:48:14.626000  
...         ...          ...        ...        ...              ...  
3542        Mar            7          3 2025-03-23  10:34:54.894000  
3543        Mar            7          3 2025-03-23  14:43:37.362000  
3544        Mar            7          3 2025-03-23  14:44:16.864000  
3545        Mar            7          3 2025-03-23  15:47:28.723000  
3546        Mar            7          3 2025-03-23  18:11:38.635000  

[3547 rows x 11 columns]

# fixing the time to be more readable

In [10]:
df["before_semicolon"] = df["Time"].str.split(":").str[0:2]

In [11]:
df

hour_of_day cash_type  money    coffee_name Time_of_Day Weekday  \
0              10      card  38.70          Latte     Morning     Fri   
1              12      card  38.70  Hot Chocolate   Afternoon     Fri   
2              12      card  38.70  Hot Chocolate   Afternoon     Fri   
3              13      card  28.90      Americano   Afternoon     Fri   
4              13      card  38.70          Latte   Afternoon     Fri   
...           ...       ...    ...            ...         ...     ...   
3542           10      card  35.76     Cappuccino     Morning     Sun   
3543           14      card  35.76          Cocoa   Afternoon     Sun   
3544           14      card  35.76          Cocoa   Afternoon     Sun   
3545           15      card  25.96      Americano   Afternoon     Sun   
3546           18      card  35.76          Latte       Night     Sun   

     Month_name  Weekdaysort  Monthsort       Date             Time  \
0           Mar            5          3 2024-03-01  10:15:50.520000   
1           Mar            5          3 2024-03-01  12:19:22.539000   
2           Mar            5          3 2024-03-01  12:20:18.089000   
3           Mar            5          3 2024-03-01  13:46:33.006000   
4           Mar            5          3 2024-03-01  13:48:14.626000   
...         ...          ...        ...        ...              ...   
3542        Mar            7          3 2025-03-23  10:34:54.894000   
3543        Mar            7          3 2025-03-23  14:43:37.362000   
3544        Mar            7          3 2025-03-23  14:44:16.864000   
3545        Mar            7          3 2025-03-23  15:47:28.723000   
3546        Mar            7          3 2025-03-23  18:11:38.635000   

     before_semicolon  
0            [10, 15]  
1            [12, 19]  
2            [12, 20]  
3            [13, 46]  
4            [13, 48]  
...               ...  
3542         [10, 34]  
3543         [14, 43]  
3544         [14, 44]  
3545         [15, 47]  
3546         [18, 11]  

[3547 rows x 12 columns]

In [12]:
def x(l):
    s = l[0]+ ":" + l[1]
    return s

In [13]:
df['Time2'] = df['before_semicolon'].apply(x)

In [14]:
df.head()

hour_of_day cash_type  money    coffee_name Time_of_Day Weekday Month_name  \
0           10      card   38.7          Latte     Morning     Fri        Mar   
1           12      card   38.7  Hot Chocolate   Afternoon     Fri        Mar   
2           12      card   38.7  Hot Chocolate   Afternoon     Fri        Mar   
3           13      card   28.9      Americano   Afternoon     Fri        Mar   
4           13      card   38.7          Latte   Afternoon     Fri        Mar   

   Weekdaysort  Monthsort       Date             Time before_semicolon  Time2  
0            5          3 2024-03-01  10:15:50.520000         [10, 15]  10:15  
1            5          3 2024-03-01  12:19:22.539000         [12, 19]  12:19  
2            5          3 2024-03-01  12:20:18.089000         [12, 20]  12:20  
3            5          3 2024-03-01  13:46:33.006000         [13, 46]  13:46  
4            5          3 2024-03-01  13:48:14.626000         [13, 48]  13:48

In [15]:
df.drop(columns=["before_semicolon"],inplace=True)


In [16]:
df.drop(columns=["Time"],inplace=True)

In [17]:
df.rename(columns={"Time2" : "Time"}, inplace=True)


In [18]:
df

hour_of_day cash_type  money    coffee_name Time_of_Day Weekday  \
0              10      card  38.70          Latte     Morning     Fri   
1              12      card  38.70  Hot Chocolate   Afternoon     Fri   
2              12      card  38.70  Hot Chocolate   Afternoon     Fri   
3              13      card  28.90      Americano   Afternoon     Fri   
4              13      card  38.70          Latte   Afternoon     Fri   
...           ...       ...    ...            ...         ...     ...   
3542           10      card  35.76     Cappuccino     Morning     Sun   
3543           14      card  35.76          Cocoa   Afternoon     Sun   
3544           14      card  35.76          Cocoa   Afternoon     Sun   
3545           15      card  25.96      Americano   Afternoon     Sun   
3546           18      card  35.76          Latte       Night     Sun   

     Month_name  Weekdaysort  Monthsort       Date   Time  
0           Mar            5          3 2024-03-01  10:15  
1           Mar            5          3 2024-03-01  12:19  
2           Mar            5          3 2024-03-01  12:20  
3           Mar            5          3 2024-03-01  13:46  
4           Mar            5          3 2024-03-01  13:48  
...         ...          ...        ...        ...    ...  
3542        Mar            7          3 2025-03-23  10:34  
3543        Mar            7          3 2025-03-23  14:43  
3544        Mar            7          3 2025-03-23  14:44  
3545        Mar            7          3 2025-03-23  15:47  
3546        Mar            7          3 2025-03-23  18:11  

[3547 rows x 11 columns]

# see the most purchased item in which they earned a lot

In [19]:
df2=df.groupby("coffee_name")

In [20]:
df2["money"].sum().sort_values(ascending= False)

coffee_name
Latte                  26875.30
Americano with Milk    24751.12
Cappuccino             17439.14
Americano              14650.26
Hot Chocolate           9933.46
Cocoa                   8521.16
Cortado                 7384.86
Espresso                2690.28
Name: money, dtype: float64

# check if there is any other cash type exsist

In [21]:
df["cash_type"].unique()

array(['card'], dtype=object)

In [22]:
df4=df.groupby("coffee_name")

# check the most time of day in which they buy something

In [23]:
df.describe(include="O")

cash_type          coffee_name Time_of_Day Weekday Month_name   Time
count       3547                 3547        3547    3547       3547   3547
unique         1                    8           3       7         12    915
top         card  Americano with Milk   Afternoon     Tue        Mar  10:12
freq        3547                  809        1205     572        494     12

# check the variety of prices in each item

In [24]:
Caffe_Category = df.pivot_table(index="coffee_name", values="money",aggfunc=["min", "max"])
Caffe_Category

min   max
                     money money
coffee_name                     
Americano            23.02  28.9
Americano with Milk  27.92  33.8
Cappuccino           32.82  38.7
Cocoa                32.82  38.7
Cortado              23.02  28.9
Espresso             18.12  24.0
Hot Chocolate        32.82  38.7
Latte                32.82  38.7

# check the amount of money increased on each item

In [25]:
Caffe_Category["difference"] = Caffe_Category[("max", "money")] - Caffe_Category[("min", "money")]
Caffe_Category


min   max difference
                     money money           
coffee_name                                
Americano            23.02  28.9       5.88
Americano with Milk  27.92  33.8       5.88
Cappuccino           32.82  38.7       5.88
Cocoa                32.82  38.7       5.88
Cortado              23.02  28.9       5.88
Espresso             18.12  24.0       5.88
Hot Chocolate        32.82  38.7       5.88
Latte                32.82  38.7       5.88

In [26]:
df

hour_of_day cash_type  money    coffee_name Time_of_Day Weekday  \
0              10      card  38.70          Latte     Morning     Fri   
1              12      card  38.70  Hot Chocolate   Afternoon     Fri   
2              12      card  38.70  Hot Chocolate   Afternoon     Fri   
3              13      card  28.90      Americano   Afternoon     Fri   
4              13      card  38.70          Latte   Afternoon     Fri   
...           ...       ...    ...            ...         ...     ...   
3542           10      card  35.76     Cappuccino     Morning     Sun   
3543           14      card  35.76          Cocoa   Afternoon     Sun   
3544           14      card  35.76          Cocoa   Afternoon     Sun   
3545           15      card  25.96      Americano   Afternoon     Sun   
3546           18      card  35.76          Latte       Night     Sun   

     Month_name  Weekdaysort  Monthsort       Date   Time  
0           Mar            5          3 2024-03-01  10:15  
1           Mar            5          3 2024-03-01  12:19  
2           Mar            5          3 2024-03-01  12:20  
3           Mar            5          3 2024-03-01  13:46  
4           Mar            5          3 2024-03-01  13:48  
...         ...          ...        ...        ...    ...  
3542        Mar            7          3 2025-03-23  10:34  
3543        Mar            7          3 2025-03-23  14:43  
3544        Mar            7          3 2025-03-23  14:44  
3545        Mar            7          3 2025-03-23  15:47  
3546        Mar            7          3 2025-03-23  18:11  

[3547 rows x 11 columns]

# check the most month in which there is high money

In [27]:
df["Monthsort"].max()

12

In [28]:
df.describe()

hour_of_day        money  Weekdaysort    Monthsort  \
count  3547.000000  3547.000000  3547.000000  3547.000000   
mean     14.185791    31.645216     3.845785     6.453905   
min       6.000000    18.120000     1.000000     1.000000   
25%      10.000000    27.920000     2.000000     3.000000   
50%      14.000000    32.820000     4.000000     7.000000   
75%      18.000000    35.760000     6.000000    10.000000   
max      22.000000    38.700000     7.000000    12.000000   
std       4.234010     4.877754     1.971501     3.500754   

                                Date  
count                           3547  
mean   2024-10-04 17:34:43.676346368  
min              2024-03-01 00:00:00  
25%              2024-07-17 12:00:00  
50%              2024-10-10 00:00:00  
75%              2025-01-11 00:00:00  
max              2025-03-23 00:00:00  
std                              NaN

In [29]:
Month= df.groupby("Monthsort") 

In [30]:
Month["money"].sum().sort_values(ascending=False)

Monthsort
3     15891.64
10    13891.16
2     13215.48
9      9988.64
11     8590.54
12     8237.74
5      8164.42
6      7617.76
8      7613.84
7      6915.94
1      6398.86
4      5719.56
Name: money, dtype: float64

# check the most day in which there is high money

In [31]:
Day= df.groupby("Weekdaysort")
Day["money"].sum().sort_values(ascending=False)

Weekdaysort
2    18168.38
1    17363.10
5    16802.66
4    16091.40
3    15750.46
6    14733.52
7    13336.06
Name: money, dtype: float64

In [32]:
df["money"].sum()

np.float64(112245.57999999999)

# check which part of the day they earn more

In [33]:
TimeDay_money = df.pivot_table(index="Time_of_Day", values="money",aggfunc=["sum"])
TimeDay_money

sum
                money
Time_of_Day          
Afternoon    38130.04
Morning      35929.20
Night        38186.34

# check at which time they earn a lot

In [34]:
Time_money = df.pivot_table(index="Time", values="money",aggfunc=["sum"])
Time_money

sum
       money
Time        
06:50  35.76
06:52  30.86
06:54  56.82
06:59  25.96
07:01  61.72
...      ...
22:54  61.72
22:55  27.92
22:56  32.82
22:58  32.82
22:59  32.82

[915 rows x 1 columns]

In [35]:
Time_money.describe()

sum
            money
count  915.000000
mean   122.672765
std     63.534226
min     21.060000
25%     71.520000
50%    115.600000
75%    163.120000
max    392.860000

In [36]:
row = Time_money[Time_money[("sum","money")] == 392.86]
print(row)

          sum
        money
Time         
10:12  392.86


# check at what month they earn a lot

In [37]:
x = df.pivot_table(
    index=pd.Grouper(freq='M', key='Date'),
    values='money',      
    aggfunc='sum'
)
x

C:\Users\Marko\AppData\Local\Temp\ipykernel_8152\3142767263.py:2: FutureWarning: 'M' is deprecated and will be removed in a future version, please use 'ME' instead.
  index=pd.Grouper(freq='M', key='Date'),


money
Date                
2024-03-31   5905.20
2024-04-30   5719.56
2024-05-31   8164.42
2024-06-30   7617.76
2024-07-31   6915.94
2024-08-31   7613.84
2024-09-30   9988.64
2024-10-31  13891.16
2024-11-30   8590.54
2024-12-31   8237.74
2025-01-31   6398.86
2025-02-28  13215.48
2025-03-31   9986.44

In [38]:
x.describe()

money
count     13.000000
mean    8634.275385
std     2559.827631
min     5719.560000
25%     6915.940000
50%     8164.420000
75%     9986.440000
max    13891.160000

In [39]:
with pd.ExcelWriter('stocks_weather.xlsx') as writer:
    x.to_excel(writer, sheet_name="Pro-Month",index=False)
    Time_money.to_excel(writer, sheet_name="Tim_mon",index=False)
    TimeDay_money.to_excel(writer, sheet_name="Tim_Day_mon",index=False)
    Caffe_Category.to_excel(writer, sheet_name="Caff_categ",index=False)

NotImplementedError: Writing to Excel with MultiIndex columns and no index ('index'=False) is not yet implemented.

# Coffee Shop Sales Summary

- Latte is the most purchased item, accounting for 24% of total sales.
  
- There is only one payment type, which is card.

- The afternoon is the time of day when the café earns the most, with 34% of total sales.

- All items have increased by 5.88%.

- In March, the café’s earnings increased by 14%, while in April, they decreased by 5%.

- On Tuesday, earnings were the highest at 16%, while on Sunday, they were the lowest at 11%.

- During the afternoon, morning, and night, the café earns roughly 34% in each period.

- At 10:12, there are a lot of customer requests.

- From 2025-02-28 to 2025-03-31, earnings increased by 11%, and from 2024-10-31 to 2024-11-30, by 12%.

In [40]:
df

hour_of_day cash_type  money    coffee_name Time_of_Day Weekday  \
0              10      card  38.70          Latte     Morning     Fri   
1              12      card  38.70  Hot Chocolate   Afternoon     Fri   
2              12      card  38.70  Hot Chocolate   Afternoon     Fri   
3              13      card  28.90      Americano   Afternoon     Fri   
4              13      card  38.70          Latte   Afternoon     Fri   
...           ...       ...    ...            ...         ...     ...   
3542           10      card  35.76     Cappuccino     Morning     Sun   
3543           14      card  35.76          Cocoa   Afternoon     Sun   
3544           14      card  35.76          Cocoa   Afternoon     Sun   
3545           15      card  25.96      Americano   Afternoon     Sun   
3546           18      card  35.76          Latte       Night     Sun   

     Month_name  Weekdaysort  Monthsort       Date   Time  
0           Mar            5          3 2024-03-01  10:15  
1           Mar            5          3 2024-03-01  12:19  
2           Mar            5          3 2024-03-01  12:20  
3           Mar            5          3 2024-03-01  13:46  
4           Mar            5          3 2024-03-01  13:48  
...         ...          ...        ...        ...    ...  
3542        Mar            7          3 2025-03-23  10:34  
3543        Mar            7          3 2025-03-23  14:43  
3544        Mar            7          3 2025-03-23  14:44  
3545        Mar            7          3 2025-03-23  15:47  
3546        Mar            7          3 2025-03-23  18:11  

[3547 rows x 11 columns]